codeowners/src/lib.rs

```diff
@@ -1,455 +1,28 @@
-//! files
-//!
-//! # Examples
-//!
-//! Typical use involves resolving a CODEOWNERS file, parsing it,
-//! then querying target paths
-//!
-//! ```no_run
-//! extern crate codeowners;
-//! use std::env;
-//!
-//! fn main() {
-//!   if let (Some(owners_file), Some(path)) =
-//!      (env::args().nth(1), env::args().nth(2)) {
-//!      let owners = codeowners::from_path(owners_file).unwrap();
-//!      match owners.of(&path) {
-//!        None => println!("{} is up for adoption", path),
-//!        Some(owners) => {
-//!           for owner in owners {
-//!             println!("{}", owner);
-//!           }
-//!        }
-//!      }
-//!   }
-//! }
-//! ```
-#![allow(missing_docs)]
+mod github;
 
-#[cfg(test)]
-#[macro_use]
-extern crate pretty_assertions;
+use std::{io::Read, path::Path};
 
-use glob::Pattern;
-use lazy_static::lazy_static;
-use regex::Regex;
-use std::{
-    fmt,
-    fs::File,
-    io::{BufRead, BufReader, Read},
-    path::{Path, PathBuf},
-    str::FromStr,
-};
+pub use github::{GitHubOwner, GitHubOwners};
 
-const CODEOWNERS: &str = "CODEOWNERS";
+pub trait OwnersOfPath {
+    type Owner;
 
-/// Various types of owners
-///
-/// Owners supports parsing from strings as well as displaying as strings
-///
-/// # Examples
-///
-/// ```rust
-/// let raw = "@org/team";
-/// assert_eq!(
-///   raw.parse::<codeowners::Owner>().unwrap().to_string(),
-///   raw
-/// );
-/// ```
-#[derive(Debug, PartialEq, Clone)]
-pub enum Owner {
-    /// Owner in the form @username
-    Username(String),
-    /// Owner in the form @org/Team
-    Team(String),
-    /// Owner in the form user@domain.com
-    Email(String),
-}
-
-impl fmt::Display for Owner {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        let inner = match *self {
-            Owner::Username(ref u) => u,
-            Owner::Team(ref t) => t,
-            Owner::Email(ref e) => e,
-        };
-        f.write_str(inner.as_str())
-    }
-}
-
-impl FromStr for Owner {
-    type Err = String;
-
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        lazy_static! {
-            static ref TEAM: Regex = Regex::new(r"^@\S+/\S+").unwrap();
-            static ref USERNAME: Regex = Regex::new(r"^@\S+").unwrap();
-            static ref EMAIL: Regex = Regex::new(r"^\S+@\S+").unwrap();
-        }
-        if TEAM.is_match(s) {
-            Ok(Owner::Team(s.into()))
-        } else if USERNAME.is_match(s) {
-            Ok(Owner::Username(s.into()))
-        } else if EMAIL.is_match(s) {
-            Ok(Owner::Email(s.into()))
-        } else {
-            Err(String::from("not an owner"))
-        }
-    }
-}
-
-/// Mappings of owners to path patterns
-#[derive(Debug, PartialEq, Clone)]
-pub struct Owners {
-    paths: Vec<(Pattern, Vec<Owner>)>,
-}
-
-impl Owners {
     /// Resolve a list of owners matching a given path
-    pub fn of<P>(&self, path: P) -> Option<&Vec<Owner>>
+    fn of<P>(&self, path: P) -> Option<Vec<Self::Owner>>
```

Comment:
removed this reference in favor of an owned `Vec` since it makes the interface more flexible to implement for each CODEOWNERS provider